vscode/src/services/TreeViewProvider.ts

```diff
@@ -28,33 +37,51 @@ export class TreeViewProvider implements vscode.TreeDataProvider<vscode.TreeItem
     }
 
     /**
-     * Updates the tree view with the provided tree items.
+     * Updates the tree view with the provided tree items, filtering out any
+     * that do not meet the required criteria to show.
      */
-    public updateTree(treeItems: CodySidebarTreeItem[]): void {
+    public async updateTree(treeItems: CodySidebarTreeItem[]): Promise<void> {
+        this.treeItems = treeItems
+        return this.refresh()
+    }
+
+    /**
+     * Refreshes the visible tree items, filtering out any
+     * that do not meet the required criteria to show.
+     */
+    public async refresh(): Promise<void> {
+        // TODO(dantup): This method can be made not-async again when we don't need to call evaluateFeatureFlag
         const updatedTree: vscode.TreeItem[] = []
-        treeItems.forEach(item => {
+        this.treeNodes = updatedTree // Set this before any awaits so last call here always wins regardless of async scheduling.
+        for (const item of this.treeItems) {
+            if (item.requireFeature && !(await this.featureFlagProvider.evaluateFeatureFlag(item.requireFeature))) {
```

Comment:
```suggestion
            if (this.type === 'support') {
                if (item.requireFeature && !(await this.featureFlagProvider.evaluateFeatureFlag(item.requireFeature))) {
                    continue
                }

                if (item.requireUpgradeAvailable && !(this.authStatus?.userCanUpgrade ?? false)) {
                    console.log(this.authStatus)
                    continue
                }
            }
```

Should we run this for the support sidebar only so the chat sidebar which gets refreshed a lot won't make new requests to the featureFlagProvider on each refresh?

Reply:
I don't think `this.type === 'support'` matters, because the other kinds won't have `item.requireFeature` set, so it won't call `evaluateFeatureFlag`?

We could check for dotcom, but this code was intended to be fairly generic and not specific to things that require dotcom (maybe if we want to do that, we should add a `requiresDotCom` or similar flag to `item`?)

(our of interest, how heavily are feature flags cached? I assumed in most cases this would be immediately returned without any network request)

Reply:
Yes that looks good to me. Sorry for the late reply!